L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -0,0 +1,266 @@
+void HGCalClusteringImpl::clusterizeDR( const edm::PtrVector<l1t::HGCalTriggerCell> & triggerCellsPtrs, 
+                                      l1t::HGCalClusterBxCollection & clusters
+    ){
+
+    bool isSeed[triggerCellsPtrs.size()];
+    
+    /* search for cluster seeds */
+    int itc(0);
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc,++itc ){
+        isSeed[itc] = ( (*tc)->mipPt() > seedThreshold_) ? true : false;
+    }
+    
+    /* clustering the TCs */
+    std::vector<l1t::HGCalCluster> clustersTmp;
+
+    itc=0;
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc,++itc ){
+            
+        if( (*tc)->mipPt() < triggerCellThreshold_ ){
+            continue;
+        }
+        
+        /* searching for TC near the center of the cluster  */
+        int iclu=0;
+        vector<int> tcPertinentClusters; 
+        for( std::vector<l1t::HGCalCluster>::iterator clu = clustersTmp.begin(); clu != clustersTmp.end(); ++clu,++iclu ){
+            if( this->isPertinent(**tc, *clu, dr_) ){
+                tcPertinentClusters.push_back(iclu);
+            }
+        }
+        if( tcPertinentClusters.size() == 0 && isSeed[itc] ){
+            clustersTmp.emplace_back( *tc );
+        }
+        else if ( tcPertinentClusters.size() > 0 ){
+         
+            unsigned minDist(300);
+            unsigned targetClu(0);
+                        
+            for( std::vector<int>::const_iterator iclu = tcPertinentClusters.begin(); iclu != tcPertinentClusters.end(); ++iclu ){
+                double d = clustersTmp.at(*iclu).distance(**tc);
+                if( d < minDist ){
+                    minDist = d;
+                    targetClu = *iclu;
+                }
+            } 
+
+            clustersTmp.at(targetClu).addTriggerCell( *tc );                    
+
+        }
+    }
+
+    /* store clusters in the persistent collection */
+    for( unsigned i(0); i<clustersTmp.size(); ++i ){
+        clusters.push_back( 0, clustersTmp.at(i) );
+    }
+    
+}
+
+
+
+/* NN-algorithms */
+
+/* storing trigger cells into vector per layer and per endcap */
+void HGCalClusteringImpl::triggerCellReshuffling( const edm::PtrVector<l1t::HGCalTriggerCell> & triggerCellsPtrs, 
+                                                  std::array< std::array<std::vector<edm::Ptr<l1t::HGCalTriggerCell>>,40>,2> & reshuffledTriggerCells 
+    ){
+
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc){
+        int endcap = (*tc)->zside() == -1 ? 0 : 1 ;
+        HGCalDetId tcDetId( (*tc)->detId() );
+        int subdet = tcDetId.subdetId();
+        int layer = -1;
+
+        if( subdet == HGCEE ){ 
+            layer = (*tc)->layer();
+        }
+        else if( subdet == HGCHEF ){
+            layer = (*tc)->layer() + 28;
```

Comment:
Make a descriptive constant for magic number 28 (or get from geometry/topology if possible)